alf/algorithms/muzero_algorithm.py

```diff
@@ -295,70 +295,152 @@ def _hook(grad, name):
+                torch.arange(T)[:, None] + torch.min(
+                    steps_to_episode_end.unsqueeze(-1),
+                    torch.arange(R + 1))  # [T, R + 1]
+            )  # [B, T, R + 1]
+
+            def _unfold1_adapting_episode_ends(x):
+                return x[capped_unfold1_index]
+
+            if self._reanalyze_ratio > 0:
+                # Here we assume state and info have similar name scheme.
+                mcts_state_field = 'state' + info_path[len('rollout_info'):]
+
+                # Applying the "unfold" trick, where we do reanalyze from the
+                # starting position of B size-T trajectory for an unroll steps
+                # of T + R - 1, and unfold it to [B, T, R + 1]
+                if self._reanalyze_ratio < 1:
+                    r = torch.randperm(B) < B * self._reanalyze_ratio
+                    # [B', T + R, ...], B' = B * reanalyze_ratio
+                    r_candidate_actions, r_candidate_action_policy, r_values = self._reanalyze(
+                        replay_buffer, start_env_ids[r], start_positions[r],
+                        mcts_state_field, T + R - 1)
+                    # [B', T, R + 1, ...]
+                    r_candidate_actions, r_candidate_action_policy, r_values = alf.nest.map_structure(
+                        _unfold1_adapting_episode_ends,
+                        (r_candidate_actions, r_candidate_action_policy,
+                         r_values))
+                else:
+                    # [B, T + R, ...]
+                    candidate_actions, candidate_action_policy, values = self._reanalyze(
+                        replay_buffer, start_env_ids, start_positions,
+                        mcts_state_field, T + R - 1)
+
+                    # [B, T, R + 1, ...]
+                    candidate_actions, candidate_action_policy, values = alf.nest.map_structure(
+                        _unfold1_adapting_episode_ends,
+                        (candidate_actions, candidate_action_policy, values))
+
             if self._reanalyze_ratio < 1:
                 if self._td_steps >= 0:
+                    # [B, T + R]
                     values = self._calc_bootstrap_return(
-                        replay_buffer, env_ids, positions, value_field)
+                        replay_buffer, folded_env_ids, folded_positions,
+                        value_field)
                 else:
+                    # [B, T + R]
                     values = self._calc_monte_carlo_return(
-                        replay_buffer, env_ids, positions, value_field)
+                        replay_buffer, folded_env_ids, folded_positions,
+                        value_field)
 
+                # [B, T, R + 1]
+                values = _unfold1_adapting_episode_ends(values)
+
+                # [B, T, R + 1]
                 candidate_actions = replay_buffer.get_field(
                     candidate_actions_field, env_ids, positions)
+                # [B, T, R + 1]
```

Comment:
candidate_actions and candidate_action_policy can be retrieved using folded_env_ids and folded_positions so that we can do a single unfolding for values, candidate_actions and candidate_action_policy as line 452 (unfolding at line 410, 421 and 438 can be removed).

Reply:
Thanks for the suggestion. Updated.
I leave `game_overs`  and `discount` below as they are since the retrieval along will not benefit from folded + unfold and `==` should be trivial enough.